File: datastruct/tree/trie/src/scala/PrefixTreeApp.scala
import scala.language.postfixOps
import Ordering.Implicits._ // for list comparison

object PrefixTreeApp {
  // lazy function to find all candidates start with given prefix
  def findAll[K, V] (t: PrefixTree.Tree[K, V],
                     key: List[K]): Stream[(List[K], V)] = {
    def enum[K, V] (cs: List[(List[K], PrefixTree.Tree[K, V])]): Stream[(List[K], V)] =
      cs.toStream.flatMap { p => mapAppend(p._1, findAll(p._2, List()))}
    def find(cs: List[(List[K], PrefixTree.Tree[K, V])],
             key: List[K]): Stream[(List[K], V)] =
      cs match {
        case List() => Stream.empty
        case p :: ps => {
          val (k, tr) = p
          if (k.startsWith(key)) {
            mapAppend(k, findAll(tr, List()))
          } else if (key.startsWith(k)) {
            mapAppend(k, findAll(tr, key.drop(k.length)))
          } else {
            find(ps, key)
          }
        }
      }
    if (key.isEmpty) {
      val ps = enum(t.children)
      if (t.value.isEmpty) ps else (List(), t.value.get) #:: ps
    } else {
      find(t.children, key)
    }
  }

  def mapAppend[A, B] (a: List[A], ps: Stream[(List[A], B)]) : Stream[(List[A], B)] =
    ps.map {p => (a ++ p._1, p._2)}

  def lookup[K, V] (t: PrefixTree.Tree[K, V], key: List[K], n: Int): List[(List[K], V)] =
    findAll(t, key).take(n).toList

  // T9 (Textonym) lookup

  val mapT9 = Map('1' -> ",.", '2' -> "abc", '3' -> "def", '4' -> "ghi", '5' -> "jkl",
                  '6' -> "mno", '7'-> "pqrs", '8' -> "tuv", '9' -> "wxyz")

  val rmapT9 = mapT9.toList.flatMap(p => p._2.toList.map {(_, p._1)}).toMap

  def digits(w: String) = w.map {rmapT9.getOrElse(_, '#')}

  def findT9[V](t: PrefixTree.Tree[Char, V], key: String): Stream[String] =
    if(key.isEmpty){
      Stream("")
    } else {
      val n = key.length
      val prefixes = t.children.toStream.filter(p => {
        val ds = digits(p._1.mkString)
        ds.startsWith(key) || key.startsWith(ds)
      })
      def find(s: String, tr: PrefixTree.Tree[Char, V]): Stream[String] =
        findT9(tr, key.drop(s.length)).map { w => (s ++ w).take(n)}
      prefixes.flatMap { p => find(p._1.mkString, p._2) }
    }

  // verification
  def testEdict() {
    val m = Map("a" -> "the first letter of English",
                "an" -> "used instead of 'a' when the following word begins with a vowel sound",
                "another" -> "one more person or thing or an extra amount",
                "abandon" -> "to leave a place, thing or person forever",
                "about" -> "on the subject of; connected with",
                "adam" -> "a character in the Bible who was the first man made by God",
                "boy" -> "a male child or, more generally, a male of any age",
                "body" -> "the whole physical structure that forms a person or animal",
                "zoo" -> "an area in which animals, especially wild animals, are kept so that people can go and look at them, or study them")
    val t = PrefixTree.fromStringList(m.toList)
    verifyLookup(m, t, "a", 5)
    verifyLookup(m, t, "a", 6)
    verifyLookup(m, t, "a", 7)
    verifyLookup(m, t, "ab", 2)
    verifyLookup(m, t, "ab", 5)
    verifyLookup(m, t, "b", 2)
    verifyLookup(m, t, "bo", 5)
    verifyLookup(m, t, "z", 3)
    println("edict verified")
  }

  def verifyLookup(m: Map[String, String], t: PrefixTree.Tree[Char, String],
                   key: String, n: Int) {
    val r = lookup(t, key.toList, n).map {p => (p._1.mkString, p._2) }
    r.foreach(p => {
      val (k, v) = p
      assert(k.startsWith(key), println(s"$k does not start with $key"))
      assert(m.contains(k), println(s"$k does not exists"))
    })
    assert(r.length <= n, println(s"expected $n results, get: " + r))
  }

  def testT9() = {
    val txt = "home good gone hood a another an"
    val words = txt.split("\\s+").toList
    val t = PrefixTree.fromString(txt)
    def norm[A <% Ordered[A]](ws: Seq[A]): Seq[A] = ws.distinct.sortWith(_ < _)
    List("4663", "22", "2668437").flatMap { _.inits.toList.init }.foreach { key =>
      val as = norm(findT9(t, key)).toList
      val bs = norm(words.map(_.take(key.length)).filter(digits(_) == key))
      assert(as == bs, println(s"$as\n!=\n$bs"))
    }
    println("t9 verified")
  }

  def test() {
    testEdict
    testT9
  }
}

File: datastruct/tree/trie/src/PrefixTreeApp.java
import java.util.*;
import java.util.stream.*;
import java.lang.Math;
import java.lang.Exception;

public class PrefixTreeApp {
    // returns up to n candidates start with given prefix
    public static <T> List<Map.Entry<String, T>> lookup(PrefixTree.Node<T> t,
                                                        String key, int n) {
        if (t == null)
            return Collections.emptyList();
        String prefix = "";
        boolean match;
        do {
            match = false;
            for (Map.Entry<String, PrefixTree.Node<T>> entry : t.subTrees.entrySet()) {
                String k = entry.getKey();
                PrefixTree.Node<T> tr = entry.getValue();
                if (k.startsWith(key)) { // key is prefix of k
                    return expand(prefix + k, tr, n);
                }
                if (key.startsWith(k)) {
                    match = true;
                    key = key.substring(k.length());
                    t = tr;
                    prefix = prefix + k;
                    break;
                }
            }
        } while (match);
        return Collections.emptyList();
    }

    static <T> List<Map.Entry<String, T>> expand(String prefix, PrefixTree.Node<T> t, int n) {
        List<Map.Entry<String, T>> res = new ArrayList<>();
        Queue<Map.Entry<String, PrefixTree.Node<T> >> q = new LinkedList<>();
        q.offer(entryOf(prefix, t));
        while(res.size() < n && !q.isEmpty()) {
            Map.Entry<String, PrefixTree.Node<T>> entry = q.poll();
            String s = entry.getKey();
            PrefixTree.Node<T> tr = entry.getValue();
            if (tr.value.isPresent()) {
                res.add(entryOf(s, tr.value.get()));
            }
            for (Map.Entry<String, PrefixTree.Node<T>> e :
                     new TreeMap<>(tr.subTrees).entrySet()) {
                q.offer(entryOf(s + e.getKey(), e.getValue()));
            }
        }
        return res;
    }

    static <K, V> Map.Entry<K, V> entryOf(K key, V val) {
        return new AbstractMap.SimpleImmutableEntry<K, V>(key, val);
    }

    // T9 map
    static final Map<Character, String> MAP_T9 = new HashMap<Character, String>(){{
            put('1', ",."); put('2', "abc"); put('3', "def");
            put('4', "ghi"); put('5', "jkl"); put('6', "mno");
            put('7', "pqrs"); put('8', "tuv"); put('9', "wxyz");
        }};

    // T9 reverse map
    static final Map<Character, Character> RMAP_T9 = new HashMap<Character, Character>(){{
            for (Character d : MAP_T9.keySet()) {
                String cs = MAP_T9.get(d);
                for (int i = 0; i < cs.length(); ++i)
                    put(cs.charAt(i), d);
            }
        }};

    /*
     * The T9 reverse map can be built with stream, but it's hard to read
     *
     *    static final Map<Character, Character> RMAP_T9 = MAP_T9.entrySet().stream()
     *        .flatMap(e -> e.getValue().chars().mapToObj(i -> (char)i)
     *                 .map(c -> entryOf(c, e.getKey())))
     *        .collect(Collectors.toMap(e -> e.getKey(), e -> e.getValue()));
     */

    public static String digits(String w) {
        StringBuilder d = new StringBuilder();
        for(int i = 0; i < w.length(); ++i)
            d.append(RMAP_T9.get(w.charAt(i)));
        return d.toString();
    }

    static class Tuple<T> {
        String prefix;
        String key;
        PrefixTree.Node<T> tree;
        Tuple(String p, String k, PrefixTree.Node<T> t) {
            prefix = p; key = k; tree = t;
        }

        public static <T> Tuple<T> of(String prefix, String key,
                                      PrefixTree.Node<T> tree) {
            return new Tuple<T>(prefix, key, tree);
        }
    }

    static String limit(int n, String s) {
        return s.substring(0, Math.min(n, s.length()));
    }

    public static <T> List<String> lookupT9(PrefixTree.Node<T> t, String key) {
        List<String> res = new ArrayList<>();
        if (t == null || key.isEmpty())
            return res;
        Queue<Tuple<T>> q = new LinkedList<>();
        q.offer(Tuple.of("", key, t));
        while (!q.isEmpty()) {
            Tuple<T> elem = q.poll();
            for (Map.Entry<String, PrefixTree.Node<T>> e :
                     elem.tree.subTrees.entrySet()) {
                String k = e.getKey();
                String ds = digits(k);
                if (ds.startsWith(elem.key)) {
                    res.add(limit(key.length(), elem.prefix + k));
                } else if (elem.key.startsWith(ds)) {
                    q.offer(Tuple.of(elem.prefix + k,
                                     elem.key.substring(k.length()),
                                     e.getValue()));
                }
            }
        }
        return res;
    }

    public static class Test {
        final static String[] testKeys =
            new String[]{"a", "an", "another", "abandon", "about", "adam", "boy", "body", "zoo"};
        final static String[] testVals =
            new String[]{"the first letter of English",
                         "used instead of 'a' when the following word begins witha vowel sound",
                         "one more person or thing or an extra amount",
                         "to leave a place, thing or person forever",
                         "on the subject of; connected with",
                         "a character in the Bible who was the first man made by God",
                         "a male child or, more generally, a male of any age",
                         "the whole physical structure that forms a person or animal",
                         "an area in which animals, especially wild animals, are kept so that people can go and look at them, or study them"};

        static void testEdict() {
            PrefixTree.Node<String> t = null;
            Map<String, String> m = new HashMap<>();
            int n = Math.min(testKeys.length, testVals.length);
            for (int i = 0; i < n; ++i) {
                t = PrefixTree.insert(t, testKeys[i], testVals[i]);
                m.put(testKeys[i], testVals[i]);
            }
            verifyLookup(m, t, "a", 5);
            verifyLookup(m, t, "a", 6);
            verifyLookup(m, t, "a", 7);
            verifyLookup(m, t, "ab", 2);
            verifyLookup(m, t, "ab", 5);
            verifyLookup(m, t, "b", 2);
            verifyLookup(m, t, "bo", 5);
            verifyLookup(m, t, "z", 3);
        }

        static void verifyLookup(Map<String, String> m, PrefixTree.Node<String> t, String key, int n) {
            System.out.format("lookup %s with limit: %d\n", key, n);
            SortedMap<String, String> m1 = new TreeMap<>();
            for (Map.Entry<String, String> e : lookup(t, key, n)) {
                m1.put(e.getKey(), e.getValue());
            }
            SortedMap<String, String> m2 =
                take(n, toSortedMap(m.entrySet().stream()
                                    .filter(e -> e.getKey().startsWith(key))));
            if (!m2.equals(m1))
                throw new RuntimeException("\n" + m1.toString() + "\n!=\n" + m2.toString());

            System.out.println("result:\n" + m1.toString());
        }

        static <T> SortedMap<String, T> take(int n, SortedMap<String, T> m) {
            return toSortedMap(m.entrySet().stream().limit(n));
        }

        static <K, V> SortedMap<K, V> toSortedMap(Stream<Map.Entry<K, V>> s) {
            return new TreeMap<>(s.collect(Collectors.toMap(e -> e.getKey(), e ->e.getValue())));
        }

        public static void testT9() {
            //System.out.println("T9 map: " + MAP_T9);
            //System.out.println("reverse T9 map: " + RMAP_T9);
            final String txt = "home good gone hood a another an";
            final String[] words = txt.split("\\W+");
            PrefixTree.Node<Integer> t = PrefixTree.fromString(txt);
            for (String testDigits : new String[]{"4663", "22", "2668437"}) {
                for (int i = 1; i <= testDigits.length(); ++i) {
                    String ds = limit(i, testDigits);
                    SortedSet<String> as = new TreeSet<>(lookupT9(t, ds));
                    SortedSet<String> bs = new TreeSet<>();
                    for (String w : words) {
                        String candidate = limit(i, w);
                        if (digits(candidate).equals(ds))
                            bs.add(candidate);
                    }
                    //System.out.println("T9 look up: " + as);
                    //System.out.println("Brute force:" + bs);
                    if (!as.equals(bs))
                        throw new RuntimeException("T9 look up " + as + "\n!=\n" +
                                                   "Brute force" + bs + "\n");
                }
            }
            System.out.println("T9 verified");
        }

        public static void test() {
            testEdict();
            testT9();
        }
    }
}
